python/lsst/ts/m2com/mock/mock_control_closed_loop.py

```diff
@@ -312,29 +313,167 @@ def load_file_cell_geometry(self, filepath):
+        distributes the force of axial hardpoints to other axial actuators.
+        It is the same idea for tangential actuators with z-moment.
 
         Parameters
         ----------
-        filepath : `pathlib.PosixPath`
-            File path of hardpoint compensation.
-        skiprows : `int`, optional
-            Number of lines to skip (int) at the start of the file. (the
-            default is 7)
+        location_axial_actuator : `list [list]`
+            Location of the axial actuators: (x, y). This should be a 72 x 2
+            matrix.
+        harpoints_axial : `list`
+            Three axial hardpoints. The order is from low to high,
+            e.g. [5, 15, 25].
+        harpoints_tangent : `list`
+            Three tangential hardpoints. This can only be [72, 74, 76] or
+            [73, 75, 77]. The order is from low to high.
+
+        Returns
+        -------
+        hd_comp_axial : `numpy.ndarray`
+            Axial hardpoint compensation matrix.
+        hd_comp_tangent : `numpy.ndarray`
+            Tangential hardpoint compensation matrix.
+
+        Raises
+        ------
+        `ValueError`
+            If the axial hardpoints are bad.
+        `ValueError`
+            If the tangential hardpoints are wrong.
         """
 
-        dataframe = pd.read_csv(filepath, skiprows=skiprows)
-        data = np.array(dataframe.iloc[:, 0])
+        # Axial hardpoints
 
-        # There are 3 axial actuators to be hardpoints
-        hd_comp_axial = data.reshape(NUM_ACTUATOR - NUM_TANGENT_LINK - 3, 3)
+        # Check the hardpoints by comparing with the expectation
+        if (
+            MockControlClosedLoop.select_axial_hardpoints(
+                location_axial_actuator, harpoints_axial[0]
+            )
+            != harpoints_axial
+        ):
+            raise ValueError("Bad selection of axial hardpoints.")
+
+        num_axial_actuators = NUM_ACTUATOR - NUM_TANGENT_LINK
+        active_actuators_axial = [
+            idx for idx in range(num_axial_actuators) if idx not in harpoints_axial
+        ]
+
+        num_hardpoints_axial = len(harpoints_axial)
 
-        hd_comp_tangent = np.array(
-            [[2 / 3, -1 / 3, 2 / 3], [2 / 3, 2 / 3, -1 / 3], [-1 / 3, 2 / 3, 2 / 3]]
+        location_axial_actuator = np.array(location_axial_actuator)
+        matrix_hp = np.append(
+            location_axial_actuator[harpoints_axial, :],
+            np.ones((num_hardpoints_axial, 1)),
+            axis=1,
         )
+        matrix_nhp = np.append(
+            location_axial_actuator[active_actuators_axial, :],
+            np.ones((num_axial_actuators - num_hardpoints_axial, 1)),
+            axis=1,
+        )
+        hd_comp_axial = matrix_nhp.dot(np.linalg.pinv(matrix_hp))
 
-        self._hd_comp = block_diag(hd_comp_axial, hd_comp_tangent)
+        # Tangential hardpoints
+        option_one = [72, 74, 76]
+        option_two = [73, 75, 77]
+
+        if harpoints_tangent == option_one:
```

Comment:
This is because I have the `self.hardpoints = [5, 15, 25, 73, 75, 77]` in `MockControlClosedLoop` already.

In addition, the code here is for my future's translation to LabVIEW or C++ (you know we are rewriting the LabVIEW code now). None of them supports the enum value to be `list`. I would prefer to be simple here to make the translation to be straightforward in a latter time. Thanks!

Reply:
I have the following in the doc string as well:

```
        hardpoints_tangent : `list`
            Three tangential hardpoints. This can only be [72, 74, 76] or
            [73, 75, 77]. The order is from low to high.
```